official/vision/beta/projects/yolo/dataloaders/yolo_input.py

```diff
@@ -0,0 +1,511 @@
+      use_tie_breaker: `boolean` indicating whether to use the anchor threshold 
+        value.
+      dtype: `str` indicating the output datatype of the datapipeline selecting 
+        from {"float32", "float16", "bfloat16"}.
+      seed: `int` the seed for random number generation. 
+    """
+    for key in masks.keys():
+      # Assert that the width and height is viable
+      assert output_size[1] % strides[str(key)] == 0
+      assert output_size[0] % strides[str(key)] == 0
+
+    # scale of each FPN level
+    self._strides = strides
+
+    # Set the width and height properly and base init:
+    self._coco91to80 = coco91to80
+    self._image_w = output_size[1]
+    self._image_h = output_size[0]
+
+    # Set the anchor boxes and masks for each scale
+    self._anchors = anchors
+    self._anchor_free_limits = anchor_free_limits
+    self._masks = {
+        key: tf.convert_to_tensor(value) for key, value in masks.items()
+    }
+    self._use_tie_breaker = use_tie_breaker
+    self._best_match_only = best_match_only
+    self._max_num_instances = max_num_instances
+
+    # Image scaling params
+    self._jitter = 0.0 if jitter is None else jitter
+    self._aug_scale_min = aug_scale_min
+    self._aug_scale_max = aug_scale_max
+    self._aug_rand_translate = aug_rand_translate
+    self._aug_rand_perspective = aug_rand_perspective
+
+    # Image spatial distortion
+    self._random_flip = random_flip
+    self._letter_box = letter_box
+    self._random_pad = random_pad
+    self._aug_rand_angle = aug_rand_angle
+
+    # Color space distortion of the image
+    self._aug_rand_saturation = aug_rand_saturation
+    self._aug_rand_brightness = aug_rand_brightness
+    self._aug_rand_hue = aug_rand_hue
+
+    # Set the per level values needed for operation
+    self._scale_xy = scale_xy
+    self._anchor_t = anchor_t
+    self._darknet = darknet
+    self._area_thresh = area_thresh
+
+    keys = list(self._masks.keys())
+
+    if self._anchor_free_limits is not None:
+      maxim = 2000
+      self._scale_up = {key: maxim // self._max_num_instances for key in keys}
+      self._anchor_t = -0.01
+    elif not self._darknet:
+      self._scale_up = {key: 6 - i for i, key in enumerate(keys)}
+    else:
+      self._scale_up = {key: 1 for key in keys}
+
+    self._seed = seed
+
+    # Set the data type based on input string
+    self._dtype = dtype
+
+  def _get_identity_info(self, image):
+    """Get an identity image op to pad all info vectors, this is used because 
+    graph compilation if there are a variable number of info objects in a list.
+    """
+    shape_ = tf.shape(image)
+    val = tf.stack([
+        tf.cast(shape_[:2], tf.float32),
+        tf.cast(shape_[:2], tf.float32),
+        tf.ones_like(tf.cast(shape_[:2], tf.float32)),
+        tf.zeros_like(tf.cast(shape_[:2], tf.float32)),
+    ])
+    return val
+
+  def _jitter_scale(self, image, shape, letter_box, jitter, random_pad,
```

Comment:
I need to understand the logic here, can't really follow.

Reply:
it applies a jitter to the crop and the aspect ratio, if scaling is not enabled, the scaling will be arbitrarily selected within the resize jitter function based on the crop selected. if scaling is enabled, then we pass the parameter crop only which  will crop and distort the image, then use affine warp to spatially augment, pad, scale and translate the image.

Reply:
cleaned it up